### PR TITLE
feature(i-am-tc) : Mountain Car (discrete action) environment

### PR DESCRIPTION
**Pull request by i-am-tc**
## Description

Implemented Mountain Car environment for discrete action space.

I intend for this branch to hold mountain car (discrete) environment work.

## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: dizoo/classic_control/mountain_car/envs/mtcar_env.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
+from typing import Any, List, Union, Optional
+import gym
+import numpy as np
+from ding.envs import BaseEnv, BaseEnvTimestep
+from ding.torch_utils import to_ndarray, to_list
+from ding.utils import ENV_REGISTRY
+
```
Comment (PaParaZz1) [inline dizoo/classic_control/mountain_car/envs/mtcar_env.py:7]: polish code style according to this [guideline](https://di-engine-docs.readthedocs.io/en/latest/21_code_style/index.html), only in your modified codes

File: dizoo/classic_control/mountain_car/envs/mtcar_env.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    > Source code of OpenAI's mountain car gym : https://is.gd/y1FkMT
+    > Gym documentation of mountain car : https://is.gd/29S0dt
+    > Based off DI-engine existing implementation of cartpole_env.py
+    > DI-engine's env creation conventions : https://is.gd/ZHLISj
+
+    Only __init__ , step, seed and reset are mandatory & impt.
+    The other methods are generally for convenience.
+    """
+
+    def __init__(self, cfg: dict = {}) -> None:
```
Comment (PaParaZz1) [inline dizoo/classic_control/mountain_car/envs/mtcar_env.py:24]: we use EasyDict for config

File: dizoo/classic_control/mountain_car/envs/mtcar_env.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    def seed(self, seed: int, dynamic_seed: bool = True) -> None:
+        self._seed = seed
+        self._dynamic_seed = dynamic_seed
+        np.random.seed(self._seed)
+
+    def reset(self) -> np.ndarray:
+        # Instantiate environment if not already done so
+        if not self._init_flag:
+            self._env = gym.make('MountainCar-v0')
+        self._init_flag = True
```
Comment (PaParaZz1) [inline dizoo/classic_control/mountain_car/envs/mtcar_env.py:48]: this line need to be in above if-branch

File: setup.py [hunk 4]
```diff
@@ -63,8 +63,9 @@
         'seaborn',
         'yapf==0.29.0',
         'responses~=0.12.1',
+        'MarkupSafe==2.0.1',
```
Comment (PaParaZz1) [inline setup.py:66]: You should upgrade your branch with main branch of official repo for each PR, these modification is from your last PR. (You can ignore this this time, but it is necessary to care about it in future work)

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: dizoo/classic_control/mountain_car/envs/mtcar_env.py (hunk 3 above)
Comment (i-am-tc) [reply dizoo/classic_control/mountain_car/envs/mtcar_env.py]: Ok. Edited this with 9ca9009e96cc94303d78efe6756b20891c34aa8a

File: dizoo/classic_control/mountain_car/envs/mtcar_env.py (hunk 2 above)
Comment (i-am-tc) [reply dizoo/classic_control/mountain_car/envs/mtcar_env.py:24]: Looking at the AtariEnv & BaseEnv (and referencing a few others), it looks like I should edit this to "cfg: dict" : correct? 
* if does seem like both "cfg: dict" and "cfg: EasyDict" will work. 
 
If so, then the cartpole environment would need to be edited as well as it is also of the form "cfg: dict = {}" , of which I unfortunately mirrored.

File: dizoo/classic_control/mountain_car/envs/mtcar_env.py (hunk 1 above)
Comment (i-am-tc) [reply dizoo/classic_control/mountain_car/envs/mtcar_env.py:7]: Edited with this commit 362aff40cd7d043ff5c69699ec84483e814dc674

File: dizoo/classic_control/mountain_car/envs/mtcar_env.py (hunk 2 above)
Comment (PaParaZz1) [reply dizoo/classic_control/mountain_car/envs/mtcar_env.py:24]: I will edit other env implementation, you can just focus on mountain car env

File: dizoo/classic_control/mountain_car/config/mtcar_rainbow_config.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+from easydict import EasyDict
+
+# DI-Engine uses EasyDict for configuration, by convention
+mtcar_rainbow_config = EasyDict(dict(
+    exp_name='mtcar_rainbow_seed0',
+    env=dict(
+        collector_env_num=8,
+        evaluator_env_num=5,
+        n_evaluator_episode=5,
+        stop_value=195,
```
Comment (PaParaZz1) [inline dizoo/classic_control/mountain_car/config/mtcar_rainbow_config.py:10]: modify proper stop_value

File: dizoo/classic_control/mountain_car/config/mtcar_rainbow_config.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    ),
+    policy=dict(
+        cuda=False,
+        priority=True,
+        discount_factor=0.97,
```
Comment (PaParaZz1) [inline dizoo/classic_control/mountain_car/config/mtcar_rainbow_config.py:15]: there are some fields with high priority to tune final performance:
- discount_factor: [0.97, 0.99, 0.995]
- nstep: [1, 3, 5]
- learning_rate: [1e-3, 3e-4, 1e-4] or more complex lr scheduler scheme
- update per collect: [3, 5, 10]
- buffer size: [2w, 5w, 10w]

File: dizoo/classic_control/mountain_car/envs/mtcar_env.py (hunk 2 above)
Comment (i-am-tc) [reply dizoo/classic_control/mountain_car/envs/mtcar_env.py]: Fixed with 68168fe0d0470d4f96450a16b7efd9561d0dc095

File: .gitpod.yml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# This configuration file was automatically generated by Gitpod.
```
Comment (PaParaZz1) [inline .gitpod.yml:1]: you can delete this yaml in this PR

File: dizoo/classic_control/example/mtcar_rainbow.py [hunk 8]
```diff
@@ -0,0 +1,13 @@
+from easydict import EasyDict
```
Comment (PaParaZz1) [inline dizoo/classic_control/example/mtcar_rainbow.py:1]: remove this file and update it in #471

**Review by PaParaZz1 — APPROVED** [review]

File: .gitpod.yml (hunk 7 above)
Comment (i-am-tc) [reply .gitpod.yml]: Done with [b5610b1](https://github.com/opendilab/DI-engine/pull/452/commits/b5610b13bafa220341fb373ebe51edc3b89e6371)

File: dizoo/classic_control/example/mtcar_rainbow.py (hunk 8 above)
Comment (i-am-tc) [reply dizoo/classic_control/example/mtcar_rainbow.py]: Done with [df15717](https://github.com/opendilab/DI-engine/pull/452/commits/df15717300358de4acb3aab8010eea18c0e26f6f)

File: dizoo/classic_control/mountain_car/config/mtcar_rainbow_config.py (hunk 5 above)
Comment (i-am-tc) [reply dizoo/classic_control/mountain_car/config/mtcar_rainbow_config.py:10]: Config file will be updates on branch "mtcar_rainbow" in #471. Will reflect these notes in #471 as well.

File: dizoo/classic_control/mountain_car/config/mtcar_rainbow_config.py (hunk 6 above)
Comment (i-am-tc) [reply dizoo/classic_control/mountain_car/config/mtcar_rainbow_config.py:15]: Config file will be updates on branch "mtcar_rainbow" in #471. Will reflect these notes in #471 as well.
